bundles/framework/tools.vitruv.framework.uuid/src/tools/vitruv/framework/uuid/UuidGeneratorAndResolverImpl.xtend

```
@@ -340,11 +340,12 @@ class UuidGeneratorAndResolverImpl implements UuidGeneratorAndResolver {
 	
 	def private static getResolvableUri(EObject object) {
 		val resourceUri = object.eResource.URI
+		val resourceRoot = object.eResource.firstRootEObject
 		// we cannot simply use EcoreUtil#getURI, because object’s domain might use XMI	UUIDs. Since
 		// XMI UUIDs can be different for different resource sets, we cannot use URIs with XMI UUIDs to identify objects
 		// across resource sets. Hence, we force hierarchical URIs. This assumes that the resolved object’s graph
 		// has the same topology in the resolving resource set. This assumption holds when we use this method.  
-		val fragmentPath = EcoreUtil.getRelativeURIFragmentPath(null, object)
+		val fragmentPath = EcoreUtil.getRelativeURIFragmentPath(resourceRoot, object)
```

What happens if a resource has multiple root elements? In this case we’d need to find the one that contains the `object`, wouldn’t we?

---

I thought about that as well. In general, Ecore is tied to using only a single root element and when having multiple root elements the first one is kind of "the central one".
Apart from transient states (like code generation for Xtext resources), the only situation I know in which multiple root elements are used are UML stereotypes. But, of course, we need to ensure that assigning UUIDs to stereotypes (or potentially other elements contained in further root elements) works properly. This will require some kind of search. I'll propose an according improvement.

---

It might make sense to first write a unit test for this situation, to be sure that it would really lead to a bug.

---

`getRelativeURIFragmentPath` contains this part:

```
if (internalEObject != ancestorEObject && ancestorEObject != null)
    {
      throw new IllegalArgumentException("The ancestor '" + ancestorEObject + "' is not an ancestor of '" + descendantEObject + "'");
    }
```

So we might also just not fix this situation, because we know we would see an exception should the situation ever arise (which is much better than having a silent failure).

---

I've added an according fix together with unit tests that led to a bug before.
In fact, the used fragment-based URIs were only a shortcut for resources with a single root elements (this is why the fragment is separated by `//`, as the number of the root element is contained in between (which is optional if there is only one).